### PR TITLE
lua: timestamp in millis

### LINKED ISSUE #10282: Lua: timestamp in millis/ns
*Description*:
We want to record timestamps in a Lua filter so that we know when requests and responses occur (and eventually derive the duration).
It appears Lua doesn't have a function to do this - os.time provides current time, but only at seconds, not millis or ns. os.clock provides a timestamp, but only for time spent in Lua.

Maybe something like `request_handle:timestamp`?

*Relevant Links*:
https://stackoverflow.com/questions/463101/lua-current-time-in-milliseconds

### PR DESCRIPTION
**Pull request by mindyor**
Description: adds Lua function to get timestamp in higher resolution. Resolutions are: milliseconds since epoch and nanoseconds since epoch. Resolution defaults to milliseconds since epoch.

Risk Level: small-medium?
Testing: added unit test.
Docs Changes: added documentation for function in `lua_filter.rst`
Release Notes: added release note in `version_history.rst`

Fixes #10282

### CODE REVIEW

File: source/extensions/filters/http/lua/lua_filter.cc [hunk 1]
```diff
@@ -499,6 +500,26 @@ int StreamHandleWrapper::luaVerifySignature(lua_State* state) {
   return 2;
 }
 
+int StreamHandleWrapper::luaTimestamp(lua_State* state) {
+  auto now = time_source_.systemTime().time_since_epoch();
+
+  auto time_unit = luaL_optstring(state, 2, "");
+  if (strcmp(time_unit, "") == 0 || strcmp(time_unit, "milliseconds_since_epoch") == 0) {
```
Comment (dio) [inline source/extensions/filters/http/lua/lua_filter.cc]: Can we do enum (enum in C++, table in lua) here? To accomplish that, I can think probably we can add an additional initializer function when we initialize the `lua_state_`. This initializer basically populate global state with e.g. a global table.

File: docs/root/configuration/http/http_filters/lua_filter.rst [hunk 2]
```diff
@@ -343,6 +343,17 @@ the length of the signature. *data* is the content which will be hashed. *dataLe
 The function returns a pair. If the first element is *true*, the second element will be empty
 which means signature is verified; otherwise, the second element will store the error message.
 
+timestamp()
+^^^^^^^^^^^
+
+.. code-block:: lua
+
+  timestamp = timestamp(format)
```
Comment (dio) [inline docs/root/configuration/http/http_filters/lua_filter.rst]:
```suggestion
  timestamp = handle:timestamp(format)
```

**Comment by dio** [comment]
Hum, sorry, it seems like merging "suggested" code change from github UI gives you DCO error :(. https://github.com/envoyproxy/envoy/blob/master/CONTRIBUTING.md#fixing-dco

File: source/extensions/filters/http/lua/lua_filter.cc [hunk 3]
```diff
@@ -499,6 +500,26 @@ int StreamHandleWrapper::luaVerifySignature(lua_State* state) {
   return 2;
 }
 
+int StreamHandleWrapper::luaTimestamp(lua_State* state) {
+  auto now = time_source_.systemTime().time_since_epoch();
+
+  auto time_unit = luaL_optstring(state, 2, "");
```
Comment (snowp) [inline source/extensions/filters/http/lua/lua_filter.cc]: i think you can do `absl::string_view unit_parameter = ...` and not have to use the strcmp helpers, relying on `empty()` and `operator==`

that said id be curious to see if @dio's suggestion makes things cleaner, wdyt?

File: source/extensions/filters/http/lua/lua_filter.cc (hunk 1 above)
Comment (mindyor) [reply source/extensions/filters/http/lua/lua_filter.cc]: Hm, were you thinking of initializing it here? https://github.com/mindyor/envoy/blob/master/source/extensions/filters/http/lua/lua_filter.cc#L518

`lua_newtable`/`lua_createtable` want a `lua_State`, but I only see `lua_state_` which is a `ThreadLocalState`. Is there a way to unwrap lua_State from here? I'm all ears!

File: source/extensions/filters/http/lua/lua_filter.cc (hunk 3 above)
Comment (mindyor) [reply source/extensions/filters/http/lua/lua_filter.cc]: Haven't quite figured out the enum thing, pushed `absl::string_view unit_paramter...` change in the meantime.

File: source/extensions/filters/http/lua/lua_filter.cc (hunk 1 above)
Comment (dio) [reply source/extensions/filters/http/lua/lua_filter.cc]: I put a quick hack: https://github.com/envoyproxy/envoy/compare/master...dio:lua-enum?expand=1. Sure you can hack it up to create a `registerEnum()` function too.

By having that,

I managed to log the enums, within each global "hook" function that we have:

```yaml
          - name: envoy.filters.http.lua
            typed_config:
              "@type": type.googleapis.com/envoy.config.filter.http.lua.v2.Lua
              inline_code: |
                function envoy_on_request(request_handle)
                  request_handle:logInfo(tostring(EnvoyTimestampResolution.MILLISECOND))
                  request_handle:logInfo(tostring(EnvoyTimestampResolution.NANOSECOND))
                end
                function envoy_on_response(response_handle)
                end
```

Comment (dio) [reply source/extensions/filters/http/lua/lua_filter.cc]: This is the way I register it. A bit ugly.

```c++
          [](lua_State* state) {
            lua_newtable(state);
            {
              LUA_ENUM(state, MILLISECOND, Timestamp::Resolution::Millisecond);
              LUA_ENUM(state, NANOSECOND, Timestamp::Resolution::Nanosecond);
            }
            lua_setglobal(state, "EnvoyTimestampResolution");
          },
```

Comment (mindyor) [reply source/extensions/filters/http/lua/lua_filter.cc]: Hm, I think I see what you're saying, let me give it a shot.

Comment (mindyor) [reply source/extensions/filters/http/lua/lua_filter.cc]: Pushing an implementation per your sketch (thank you for all of the detail!). I will try to come back to this PR a bit later this week (getting pulled into other things) to refactor it.

File: source/extensions/filters/http/lua/lua_filter.cc [hunk 4]
```diff
@@ -499,6 +518,29 @@ int StreamHandleWrapper::luaVerifySignature(lua_State* state) {
...
 
+int StreamHandleWrapper::luaTimestamp(lua_State* state) {
+  auto now = time_source_.systemTime().time_since_epoch();
+
+  absl::string_view unit_parameter = luaL_optstring(state, 2, "");
+
+  auto millisecond = get_field(state, "EnvoyTimestampResolution", "MILLISECOND");
+  auto nanosecond = get_field(state, "EnvoyTimestampResolution", "NANOSECOND");
+
+  if (unit_parameter.empty() || unit_parameter.compare(std::to_string(millisecond)) == 0) {
```
Comment (dio) [inline source/extensions/filters/http/lua/lua_filter.cc]: Hum, should we do `enumToInt()` here? 

```cpp
if (resolution_as_int_from_state == enumToInt(Timestamp::Resolution::Millisecond)) {
  ...
}
```

**Comment by dio** [comment]
Hum, I can take over this if it is needed @mindyor.

**Comment by mindyor** [comment]
Gah, sorry for ghosting. Any thoughts if I pick this back up again?

**Pull request closed without merging**